### ISSUE TITLE
Furthere improvement of evaluate!

### ISSUE THREAD

**Issue by shinaoka** [SpM-lab/SparseIR.jl#23]
When `dim=end`, `evaluate!` still allocates a lot of memory.

Benchmark result for commit bebe1861bfa7da2c996d9bfb53d10387edce988a.

```
using Revise
using SparseIR
using BenchmarkTools

beta = 1.0
wmax = 1000.0
basis = FiniteTempBasis(fermion, beta, wmax, 1e-7)
smpl = MatsubaraSampling(basis)

N = 1000000
in = zeros(ComplexF64, length(basis), N)
out = zeros(ComplexF64, length(smpl.sampling_points), N)
@benchmark evaluate!(out, smpl, in; dim=1)
```

<img width="605" alt="image" src="https://user-images.githubusercontent.com/15949860/170853708-9da2ca81-f824-469f-84ee-ca89d9a39e61.png">

```
N = 1000000
in = zeros(ComplexF64, N, length(basis))
out = zeros(ComplexF64, N, length(smpl.sampling_points))
@benchmark evaluate!(out, smpl, in; dim=2)
```
<img width="630" alt="image" src="https://user-images.githubusercontent.com/15949860/170853726-7e7c0b8b-cd8f-4657-ba60-702aa0733cba.png">

**Comment by shinaoka** [comment]
This originates from that `matop!` always apply to the first dimension of an array.
https://github.com/SpM-lab/SparseIR.jl/blob/main/src/sampling.jl#L188

**Comment by shinaoka** [comment]
Fixed in 06228ffb7ece8b1c3e1857d3e937773ec91535fa.
